pkg/middleware/middleware.go

```diff
@@ -67,6 +70,15 @@ func GetContextHandler(ats m.UserTokenService, remoteCache *remotecache.RemoteCa
 				ctx.Logger.Error("Failed to update last_seen_at", "error", err)
 			}
 		}
+
+		sessionCtx := context.Background()
+		// TODO not sure what to do with the error here
+		_ = sqlStore.InTransaction(sessionCtx, func(sessionCtx context.Context) error {
+			// Looks weird here but WithDbSession puts session into the context
+			ctx.DatasourceService = datasources.NewDatasourceService(sessionCtx, sqlStore)
+			c.Next()
+			return nil
+		})
```

Comment:
Only here should the transaction be committed.

Reply:
Don't think using `c.Next()` works as expected. We tried to use that for user auth tokens, basically doing some pre-processing before calling api call and then doing some post-processing after handling the api call, but it didn't work as expected. You may be more successful using the `response.Before` as we use here: 
https://github.com/grafana/grafana/blob/75c29566a619c3334f7045767b8c9ec819b6a45b/pkg/middleware/middleware.go#L241-L251

Reply:
Do you know whether there is some bug in Macaron or why doesn't it work as expected?

Reply:
Not sure, but remember there was something related to rendering static resources.